File: modules/zon/src/main/java/com/falsepattern/zigbrains/zon/highlight/ZonSyntaxHighlighter.java
/*
 * Copyright 2023-2024 FalsePattern
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.falsepattern.zigbrains.zon.highlight;

import com.falsepattern.zigbrains.zon.lexer.ZonLexerAdapter;
import com.falsepattern.zigbrains.zon.psi.ZonTypes;
import com.intellij.lexer.Lexer;
import com.intellij.openapi.editor.DefaultLanguageHighlighterColors;
import com.intellij.openapi.editor.HighlighterColors;
import com.intellij.openapi.editor.colors.TextAttributesKey;
import com.intellij.openapi.fileTypes.SyntaxHighlighterBase;
import com.intellij.psi.TokenType;
import com.intellij.psi.tree.IElementType;
import org.jetbrains.annotations.NotNull;

import java.util.HashMap;
import java.util.Map;

public class ZonSyntaxHighlighter extends SyntaxHighlighterBase {
    // @formatter:off
    public static final TextAttributesKey
        EQ       = createKey("EQ"           , DefaultLanguageHighlighterColors.OPERATION_SIGN),
        ID       = createKey("ID"           , DefaultLanguageHighlighterColors.INSTANCE_FIELD),
        COMMENT  = createKey("COMMENT"      , DefaultLanguageHighlighterColors.LINE_COMMENT  ),
        BAD_CHAR = createKey("BAD_CHARACTER", HighlighterColors.BAD_CHARACTER                ),
        STRING   = createKey("STRING"       , DefaultLanguageHighlighterColors.STRING        ),
        COMMA    = createKey("COMMA"        , DefaultLanguageHighlighterColors.COMMA         ),
        DOT      = createKey("DOT"          , DefaultLanguageHighlighterColors.DOT           ),
        BRACE    = createKey("BRACE"        , DefaultLanguageHighlighterColors.BRACES        );
    // @formatter:on

    private static final TextAttributesKey[] EMPTY_KEYS = new TextAttributesKey[0];
    private static final Map<IElementType, TextAttributesKey[]> KEYMAP = new HashMap<>();

    static {
        // @formatter:off
        addMapping(DOT     , ZonTypes.DOT);
        addMapping(COMMA   , ZonTypes.COMMA);
        addMapping(BRACE   , ZonTypes.LBRACE);
        addMapping(BRACE   , ZonTypes.RBRACE);
        addMapping(STRING  , ZonTypes.LINE_STRING, ZonTypes.STRING_LITERAL_SINGLE);
        addMapping(BAD_CHAR, TokenType.BAD_CHARACTER);
        addMapping(COMMENT , ZonTypes.COMMENT);
        addMapping(ID      , ZonTypes.ID);
        addMapping(EQ      , ZonTypes.EQ);
        // @formatter:on
    }

    private static void addMapping(TextAttributesKey key, IElementType... types) {
        for (var type : types) {
            KEYMAP.put(type, new TextAttributesKey[]{key});
        }
    }

    private static TextAttributesKey createKey(String name, TextAttributesKey fallback) {
        return TextAttributesKey.createTextAttributesKey("ZON_" + name, fallback);
    }

    @Override
    public @NotNull Lexer getHighlightingLexer() {
        return new ZonLexerAdapter();
    }

    @Override
    public TextAttributesKey @NotNull [] getTokenHighlights(IElementType tokenType) {
        if (KEYMAP.containsKey(tokenType)) {
            return KEYMAP.get(tokenType);
        }
        return EMPTY_KEYS;
    }
}

File: core/src/main/kotlin/com/falsepattern/zigbrains/zon/highlighting/ZonSyntaxHighlighter.kt
/*
 * This file is part of ZigBrains.
 *
 * Copyright (C) 2023-2025 FalsePattern
 * All Rights Reserved
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * ZigBrains is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, only version 3 of the License.
 *
 * ZigBrains is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with ZigBrains. If not, see <https://www.gnu.org/licenses/>.
 */

package com.falsepattern.zigbrains.zon.highlighting

import com.falsepattern.zigbrains.zig.highlighter.ZigSyntaxHighlighter.Companion.BAD_CHAR
import com.falsepattern.zigbrains.zig.highlighter.ZigSyntaxHighlighter.Companion.CHAR
import com.falsepattern.zigbrains.zig.highlighter.ZigSyntaxHighlighter.Companion.COMMENT
import com.falsepattern.zigbrains.zig.highlighter.ZigSyntaxHighlighter.Companion.KEYWORD
import com.falsepattern.zigbrains.zig.highlighter.ZigSyntaxHighlighter.Companion.NUMBER
import com.falsepattern.zigbrains.zig.highlighter.ZigSyntaxHighlighter.Companion.STRING
import com.falsepattern.zigbrains.zig.highlighter.ZigSyntaxHighlighter.Companion.STRING_ESC_I_C
import com.falsepattern.zigbrains.zig.highlighter.ZigSyntaxHighlighter.Companion.STRING_ESC_I_U
import com.falsepattern.zigbrains.zig.highlighter.ZigSyntaxHighlighter.Companion.STRING_ESC_V
import com.falsepattern.zigbrains.zig.psi.ZigTypes
import com.falsepattern.zigbrains.zon.lexer.ZonHighlightingLexer
import com.falsepattern.zigbrains.zon.psi.ZonTypes
import com.intellij.openapi.editor.colors.TextAttributesKey
import com.intellij.openapi.fileTypes.SyntaxHighlighterBase
import com.intellij.psi.StringEscapesTokenTypes
import com.intellij.psi.TokenType
import com.intellij.psi.tree.IElementType

class ZonSyntaxHighlighter : SyntaxHighlighterBase() {
    override fun getHighlightingLexer() = ZonHighlightingLexer()

    override fun getTokenHighlights(tokenType: IElementType?) = KEYMAP.getOrDefault(tokenType, EMPTY_KEYS)

    companion object {
        private val EMPTY_KEYS = emptyArray<TextAttributesKey>()
        private val KEYMAP = HashMap<IElementType, Array<TextAttributesKey>>()

        private fun addMapping(key: TextAttributesKey, vararg types: IElementType) = types.forEach { KEYMAP[it] = arrayOf(key) }

        init {
            // @formatter:off
            addMapping(COMMENT, ZonTypes.LINE_COMMENT)
            addMapping(
                KEYWORD,
                *ZonTypes::class.java
                    .fields
                    .filter {
                        it.name.startsWith("KEYWORD_")
                    }
                    .map { it.get(null) as IElementType }
                    .toTypedArray()
            )
            addMapping(STRING, ZonTypes.STRING_LITERAL_SINGLE, ZonTypes.STRING_LITERAL_MULTI, ZigTypes.STRING_LITERAL_SINGLE, ZigTypes.STRING_LITERAL_MULTI)
            addMapping(STRING_ESC_V, StringEscapesTokenTypes.VALID_STRING_ESCAPE_TOKEN)
            addMapping(STRING_ESC_I_C, StringEscapesTokenTypes.INVALID_CHARACTER_ESCAPE_TOKEN)
            addMapping(STRING_ESC_I_U, StringEscapesTokenTypes.INVALID_UNICODE_ESCAPE_TOKEN)
            addMapping(BAD_CHAR, TokenType.BAD_CHARACTER)
            addMapping(NUMBER, ZonTypes.INTEGER, ZonTypes.FLOAT, ZonTypes.NUM_NAN, ZonTypes.NUM_INF)
            addMapping(CHAR, ZonTypes.CHAR_LITERAL, ZigTypes.CHAR_LITERAL)

            // @formatter:on
        }
    }
}